artemis-commons/src/main/java/org/apache/activemq/artemis/utils/AbstractInterner.java

```diff
@@ -0,0 +1,157 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.activemq.artemis.utils;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.util.internal.MathUtil;
+import io.netty.util.internal.PlatformDependent;
+
+/**
+ * Thread-safe {@code <T>} interner.
+ * <p>
+ * Differently from {@link String#intern()} it contains a fixed amount of entries and
+ * when used by concurrent threads it doesn't ensure the uniqueness of the entries ie
+ * the same entry could be allocated multiple times by concurrent calls.
+ */
+public abstract class AbstractInterner<T> {
+
+   private final T[] entries;
```

Comment:
this doesn't allow GC to occur ever on any object held by this class, if the object is de-referenced.

Reply:
When the decoder will be deallocated the (few) instances will be removed or am I missing something?

Reply:
A decoder for a connection (hopefully if clients do the right thing and re-use them, should hang around a long time)

Reply:
It will have anyway a fixed (and very low) footprint, hence I don't see any problem on it

Reply:
The implementation you've shared is doing a rather different thing: so probably I need to explain better what this PR is aiming to do instead..

Reply:
If so yes you do, and you should avoid the word intern then, as typically expectation is similar behaviour to String.intern (which is what guava's does)

Reply:
fair enough: I've taken the name from here https://github.com/OpenHFT/Java-Lang/blob/master/lang/src/main/java/net/openhft/lang/pool/StringInterner.java
And I've always thought is not such a bad name

Reply:
Yeah but notice how its named under package pool, because really its just pooling the x number of last Strings that the CharSequence has been toString'd.